step functions (#18908)
DataDog/documentation: content/en/serverless/step_functions/installation.md

      text: 'Configure Serverless Monitoring'
    - link: "/integrations/amazon_lambda/"
      tag: "Documentation"
      text: "AWS Lambda Integration"
---

### Requirements
* The full Step Function execution length must be less than 15 minutes for full traces.
* Linked Lambda traces are supported for Node.js (layer v94+) and Python (layer v75+) runtimes.

### Setup

{{< tabs >}}
{{% tab "Serverless Framework" %}}

For developers using [Serverless Framework][4] to deploy serverless applications, use the Datadog Serverless Framework Plugin.

1. If you have not already, install the [Datadog Serverless Framework Plugin][1] v5.40.0+:

    ```shell
    serverless plugin install --name serverless-plugin-datadog
    ```

2. Ensure you have deployed the [Datadog Lambda Forwarder][2], a Lambda function that ships logs from AWS to Datadog, and that you are using v3.66.0+. You may need to [update your Forwarder][5].

   Take note of your Forwarder's ARN.

3. Enable all logging for your Step Function. In your AWS console, open your state machine. Click *Edit* and find the Logging section. There, set *Log level* to `ALL` and enable the *Include execution data* checkbox.

>    {{< img src="serverless/step_functions/aws_log.png" alt="AWS UI, Logging section, showing log level set to ALL." style="width:100%;" >}}

Reviewer: Sorry this change for datadog-ci and plugin were only released yesterday. Step 3 is not needed for CI and Plugin because that would be automatic for both.

<img width="942" alt="image" src="https://github.com/DataDog/documentation/assets/47579703/fd9f66bf-ddb9-427d-8110-ba912772c831">

Reviewer: - For Plugin, `enableStepFunctionTracing=true` does step 3 automatically. 
- For CI, the instrument command does this.

4. Add the following to your `serverless.yml`:

    ```yaml
    custom:
      datadog:
        site: <DATADOG_SITE>
        apiKeySecretArn: <DATADOG_API_KEY_SECRET_ARN>
        forwarderArn: <FORWARDER_ARN>

>           enableStepFunctionTracing: true

Reviewer: unindent 2 spaces and `enableStepFunctionTracing`.

     5. Toggle on *Step Functions CloudWatch Logs*. Changes take 15 minutes to take effect.

     **Note**: Log Autosubscription requires your Lambda Forwarder and Step Function to be in the same region.

   - **Manual**
     1. Ensure that your log group name has the prefix `/aws/vendedlogs/states`. 
     2. Open your AWS console and go to your Datadog Lambda Forwarder. In the *Function overview* section, click on *Add trigger*.
     3. Under *Add trigger*, in the *Trigger configuration* section, use the *Select a source* dropdown to select `CloudWatch Logs`.
     4. Under *Log group*, select the log group for your state machine. For example, `/aws/vendedlogs/states/my-state-machine`.
     5. Enter a filter name. You can choose to name it "empty filter" and leave the *Filter pattern* box blank.
4. Enable tracing on your Step Functions. Set the environment variable `DD_TRACE_ENABLED` to `true`.
5. Set up tags. Open your AWS console and go to your Step Functions state machine. Open the *Tags* section and add `env:<ENV_NAME>` and `service:<SERVICE_NAME>` tags. The `env` tag is required to see traces in Datadog. The `service` tag defaults to the state machine's name.

> 6. For Node.js and Python runtimes, you can link your Step Function traces to Lambda traces. On the Lambda Task, set the `OutputPath` and `Parameters` keys with the following: 

Reviewer: > On the Lambda Task of your Step Function's definition, set the `Parameters` key with the following:

@sumedham I am not sure about `OutputPath`. I have never tested that field. We may want to left it out until we tested that.

   ```json
   "OutputPath": "$.Payload",
   "Parameters": {
     "Payload.$": "States.JsonMerge($$, $, false)",
     "FunctionName": "${lambdaArn}"
   }
   ```

   The `JsonMerge` [intrinsic function][6] merges the [Step Functions context object][7] (`$$`) with the original Lambda's input payload (`$`). Fields of the original payload overwrite the Step Functions context object if their keys are the same.

**Example**:

{{< highlight json "hl_lines=4-7" >}}
"Lambda Read From DynamoDB": {
      "Type": "Task",
      "Resource": "arn:aws:states:::lambda:invoke",
      "OutputPath": "$.Payload",
      "Parameters": {
        "Payload.$": "States.JsonMerge($$, $, false)",
        "FunctionName": "${lambdaArn}"
      },
      "End": true

>     }

Reviewer: L136-L144 unindent by 4.

{{< highlight json "hl_lines=8-10" >}}
"Lambda Read From DynamoDB": {
      "Type": "Task",
      "Resource": "arn:aws:states:::lambda:invoke",
      "OutputPath": "$.Payload",
      "Parameters": {
        "Payload": {
          ...
          "Execution.$": "$$.Execution",
          "State.$": "$$.State",
          "StateMachine.$": "$$.StateMachine"
        },
        "FunctionName": "${lambdaArn}"
      },
      "End": true
    }
{{< /highlight >}}

If you have not yet instrumented your Lambda functions to send traces, you can [follow the steps to add the Lambda layer for your preferred runtime][3].

[1]: /logs/guide/forwarder/
[2]: /logs/guide/forwarder/?tab=cloudformation#upgrade-to-a-new-version
[3]: /logs/guide/forwarder/?tab=cloudformation#installation
[4]: /getting_started/integrations/aws/
[5]: https://app.datadoghq.com/integrations/aws
[6]: https://docs.aws.amazon.com/step-functions/latest/dg/amazon-states-language-intrinsic-functions.html#asl-intrsc-func-json-manipulate
[7]: https://docs.aws.amazon.com/step-functions/latest/dg/input-output-contextobject.html
{{% /tab %}}
{{< /tabs >}}

## See your Step Function traces in Datadog

> After you have invoked your state machine, go to the [**Trace Explorer**][2] in Datadog. Search for `service:<YOUR_STATE_MACHINE_NAME>` to see all traces associated with that state machine. 

Reviewer: Search for `service:<YOUR_STATE_MACHINE_NAME>` (if you have not set `service` tag on your state machine) ...

---
title: Install Serverless Monitoring for AWS Step Functions
kind: documentation
further_reading:
    - link: '/serverless/configuration/'
      tag: 'Documentation'
      text: 'Configure Serverless Monitoring'
    - link: "/integrations/amazon_lambda/"
      tag: "Documentation"
      text: "AWS Lambda Integration"
---

### Requirements
* The full Step Function execution length must be less than 15 minutes for full traces.
* [Linked Lambda traces][1] are supported for Node.JS (layer v94+) and Python (layer v75+) runtimes.

### Setup

{{< tabs >}}
{{% tab "Serverless Framework" %}}

> For developers using [Serverless Framework][4] to deploy their serverless applications, use the Datadog Serverless Framework Plugin.

Reviewer 2: ```suggestion
For developers using [Serverless Framework][4] to deploy serverless applications, use the Datadog Serverless Framework Plugin.
```

   {{< img src="serverless/step_functions/aws_log.png" alt="AWS UI, Logging section, showing log level set to ALL." style="width:100%;" >}}

4. Add the following to your `serverless.yml`:

    ```yaml
    custom:
      datadog:
        site: <DATADOG_SITE>
        apiKeySecretArn: <DATADOG_API_KEY_SECRET_ARN>
        forwarderArn: <FORWARDER_ARN>
          enableStepFunctionTracing: true
    ```
    - Replace `<DATADOG_SITE>` with {{< region-param key="dd_site" code="true" >}} (ensure the correct SITE is selected on the right).
    - Replace `<DATADOG_API_KEY_SECRET_ARN>` with the ARN of the AWS secret where your [Datadog API key][3] is securely stored. The key needs to be stored as a plaintext string (not a JSON blob). The `secretsmanager:GetSecretValue` permission is required. For quick testing, you can instead use `apiKey` and set the Datadog API key in plaintext.
    - Replace `<FORWARDER_ARN>` with the ARN of your Datadog Lambda Forwarder, as noted previously.

    For additional settings, see [Datadog Serverless Framework Plugin - Configuration parameters][7].

> 5. For Node.JS and Python runtimes, set `mergeStepFunctionAndLambdaTraces:true` in your `serverless.yaml` file. This links your Step Function traces with Lambda traces. If you have not instrumented your Lambda functions to send traces, you can [follow the steps to add the Lambda layer for your preferred runtime][8].

Reviewer 2: I'm seeing NodeJS or Node.js in our docs
```suggestion
5. For Node.js and Python runtimes, set `mergeStepFunctionAndLambdaTraces:true` in your `serverless.yaml` file. This links your Step Function traces with Lambda traces. If you have not instrumented your Lambda functions to send traces, you can [follow the steps to add the Lambda layer for your preferred runtime][8].
```

[1]: https://docs.datadoghq.com/serverless/libraries_integrations/plugin/
[2]: /logs/guide/forwarder/
[3]: https://app.datadoghq.com/organization-settings/api-keys
[4]: https://www.serverless.com/
[5]: /logs/guide/forwarder/?tab=cloudformation#upgrade-to-a-new-version
[6]: logs/guide/forwarder/?tab=cloudformation#installation
[7]: serverless/libraries_integrations/plugin/#configuration-parameters
[8]: /serverless/installation/#installation-instructions
{{% /tab %}}
{{% tab "Datadog Serverless CLI" %}}
1. If you have not already, install the [Datadog Serverless CLI][1] v2.18.0+.

   ```shell
   npm install -g @datadog/datadog-ci
   ```
2. Ensure you have deployed the [Datadog Lambda Forwarder][2], a Lambda function that ships logs from AWS to Datadog, and that you are using v3.66.0+. You may need to [update your Forwarder][3].

   Take note of your Forwarder's ARN.

> 3. Enable all logging for your Step Function. In your AWS console, open your state machine. Click *Edit* and find the Logging section. There, set *Log level* to `ALL` and enable the *Include execution data* checkbox.

Reviewer 2: Optional: I've seen actions bolded in other parts of the docs, example: Click **edit**, Select **option**
but there is consistency here so making an optional mention.

Author: yeah... cross-docs consistency for this type of thing is probs a style council issue (there was actually a discussion about it yesterday in slack, incidentally) — but for now i'm going to prioritize section-specific consistency.

   {{< img src="serverless/step_functions/aws_log.png" alt="AWS UI, Logging section, showing log level set to ALL." style="width:100%;" >}}
4. Instrument your Step Function.

   ```shell
   datadog-ci stepfunctions instrument --step-function <STEP_FUNCTION_ARN> --forwarder <FORWARDER_ARN>
   ```
   - Replace `<STEP_FUNCTION_ARN>` with the ARN of your Step Function.
   - Replace `<FORWARDER_ARN>` with the ARN of your Datadog Lambda Forwarder, as noted previously.

   For more information about the `datadog-ci stepfunctions` command, see the [Datadog CLI documentation][5].

> 5. For Node.JS and Python runtimes, add the flag `-- mergeStepFunctionAndLambdaTraces` in your command. This links your Step Function traces with Lambda traces. If you have not yet instrumented your Lambda functions to send traces, you can [follow the steps to add the Lambda layer for your preferred runtime][6].

Reviewer 2: ```suggestion
5. For Node.js and Python runtimes, add the flag `-- mergeStepFunctionAndLambdaTraces` in your command. This links your Step Function traces with Lambda traces. If you have not yet instrumented your Lambda functions to send traces, you can [follow the steps to add the Lambda layer for your preferred runtime][6].
```

     3. On the left, select the AWS account where your Step Function is running. Open the *Log Collection* tab.
     4. In the *Log Autosubscription* section, under *Autosubscribe Forwarder Lambda Functions*, enter the ARN of your Datadog Lambda Forwarder, as noted previously. Click *Add*.
     5. Toggle on *Step Functions CloudWatch Logs*. Changes take 15 minutes to take effect.

     **Note**: Log Autosubscription requires your Lambda Forwarder and Step Function to be in the same region.

   - **Manual**
     1. Ensure that your log group name has the prefix `/aws/vendedlogs/states`. 
     2. Open your AWS console and go to your Datadog Lambda Forwarder. In the *Function overview* section, click on *Add trigger*.
     3. Under *Add trigger*, in the *Trigger configuration* section, use the *Select a source* dropdown to select `CloudWatch Logs`.
     4. Under *Log group*, select the log group for your state machine. For example, `/aws/vendedlogs/states/my-state-machine`.
     5. Enter a filter name. You can choose to name it "empty filter" and leave the *Filter pattern* box blank.
4. Enable tracing on your Step Functions. Set the environment variable `DD_TRACE_ENABLED` to `true`.

> <!-- Where is this environment variable set? -->

Reviewer 2: Flagging leftover comment

5. Set up tags. Open your AWS console and go to your Step Functions state machine. Open the *Tags* section and add `env:<ENV_NAME>` and `service:<SERVICE_NAME>` tags. The `env` tag is required to see traces in Datadog. The `service` tag defaults to the state machine's name.

> 6. For Node.JS and Python runtimes, you can link your Step Function traces to Lambda traces. On the Lambda Task, set the `OutputPath` and `Parameters` keys with the following: 

Reviewer 2: ```suggestion
6. For Node.js and Python runtimes, you can link your Step Function traces to Lambda traces. On the Lambda Task, set the `OutputPath` and `Parameters` keys with the following: 
```

   ```json
   "OutputPath": "$.Payload",
   "Parameters": {
     "Payload.$": "States.JsonMerge($$, $, false)",
     "FunctionName": "${lambdaArn}"
   }
   ```

   The `JsonMerge` [intrinsic function][6] merges the [Step Functions context object][7] (`$$`) with the original Lambda's input payload (`$`). Fields of the original payload overwrite the Step Functions context object if their keys are the same.

   **Example**:

> {{< highlight json "hl_lines=4-7" >}}

Reviewer 2: Did you want the examples lined up under Example? As it is, it's aligned to the main page. https://a.cl.ly/RBuJJrp1

Author: yep, known issue. the `highlight` blocks don't render properly when indented.

"Lambda Read From DynamoDB": {
      "Type": "Task",
      "Resource": "arn:aws:states:::lambda:invoke",
      "OutputPath": "$.Payload",
      "Parameters": {
        "Payload.$": "States.JsonMerge($$, $, false)",
        "FunctionName": "${lambdaArn}"
      },
      "End": true
    }
{{< /highlight >}}

> Alternatively, if you have business logic defined in the payload, you could also use the following:

Reviewer 2: Is this section also supposed to be under Example?

Author: going to dedent Example so this looks less weird

{{< highlight json "hl_lines=8-10" >}}
"Lambda Read From DynamoDB": {
      "Type": "Task",
      "Resource": "arn:aws:states:::lambda:invoke",
      "OutputPath": "$.Payload",
      "Parameters": {
        "Payload": {
          ...
          "Execution.$": "$$.Execution",
          "State.$": "$$.State",
          "StateMachine.$": "$$.StateMachine"
        },
        "FunctionName": "${lambdaArn}"
      },
      "End": true
    }
{{< /highlight >}}

If you have not yet instrumented your Lambda functions to send traces, you can [follow the steps to add the Lambda layer for your preferred runtime][3].

[1]: /logs/guide/forwarder/
[2]: /logs/guide/forwarder/?tab=cloudformation#upgrade-to-a-new-version
[3]: /logs/guide/forwarder/?tab=cloudformation#installation
[4]: /getting_started/integrations/aws/
[5]: https://app.datadoghq.com/integrations/aws
[6]: https://docs.aws.amazon.com/step-functions/latest/dg/amazon-states-language-intrinsic-functions.html#asl-intrsc-func-json-manipulate
[7]: https://docs.aws.amazon.com/step-functions/latest/dg/input-output-contextobject.html
{{% /tab %}}
{{< /tabs >}}

## See your Step Function traces in Datadog

> After you have invoked your state machine, go to the [Trace Explorer][2] in Datadog. Search for `service:<YOUR_STATE_MACHINE_NAME>` to see all traces associated with that state machine. 

Reviewer 2: Optional: Here too I've also seen UI reference bolded
```suggestion
After you have invoked your state machine, go to the [**Trace Explorer**][2] in Datadog. Search for `service:<YOUR_STATE_MACHINE_NAME>` to see all traces associated with that state machine. 
```

If you cannot see your traces, see [Troubleshooting][5].

> [1]: #associated-lambda-tracing

Reviewer 2: Couldn't find this anchor link on the page, where should this link to?

---
title: Install Serverless Monitoring for AWS Step Functions
kind: documentation
further_reading:
    - link: '/serverless/configuration/'
      tag: 'Documentation'
      text: 'Configure Serverless Monitoring'
    - link: "/integrations/amazon_lambda/"
      tag: "Documentation"
      text: "AWS Lambda Integration"
---

### Requirements
* States must last less than 15 minutes for full traces.

> * [Associated Lambda tracing][1] is supported for Node.JS and Python runtimes.

Author: Will change to "Linked Lambda traces," but we as a rule don't use the word "currently" in Datadog documentation (or make any reference to any type of temporality)

### Setup

{{< tabs >}}

> {{% tab "Serverless Framework" %}}

Reviewer 3: @sumedham maybe people use both lambda and SF (quite common) would now ask should I use the forwarder for both lambda and SF? perhaps we should give them a note that we do recommend using extension for lambda while forwarder for SF?

For developers using [Serverless Framework][4] to deploy their serverless applications, use the Datadog Serverless Framework Plugin.

1. If you have not already, install the [Datadog Serverless Framework Plugin][1]:

    ```shell
    serverless plugin install --name serverless-plugin-datadog
    ```

2. Ensure you have deployed the [Datadog Lambda Forwarder][2], a Lambda function that ships logs from AWS to Datadog, and that you are using v3.66.0+. You may need to [update your Forwarder][5].

   Take note of your Forwarder's ARN.

> 3. Enable all logging for your Step Function. In your AWS console, open your state machine. Click *Edit* and find the Logging section. There, set *Log level* to `ALL` and enable the *Include execution data* checkbox.

Reviewer 4: Should we add a screenshot here?

Reviewer 4: <img width="618" alt="image" src="https://github.com/DataDog/documentation/assets/87997309/97794006-3a09-4cf9-b2cb-64b80d592361">

4. Add the following to your `serverless.yml`:

    ```yaml
    custom:
      datadog:
        site: <DATADOG_SITE>
        apiKeySecretArn: <DATADOG_API_KEY_SECRET_ARN>
        forwarderArn: <FORWARDER_ARN>

>           subscribeToStepFunctionLogs: true

Reviewer 4: ```suggestion
          enableStepFunctionTracing: true
```

Reviewer 4: We changed this recently

    ```
    - Replace `<DATADOG_SITE>` with {{< region-param key="dd_site" code="true" >}} (ensure the correct SITE is selected on the right).
    - Replace `<DATADOG_API_KEY_SECRET_ARN>` with the ARN of the AWS secret where your [Datadog API key][3] is securely stored. The key needs to be stored as a plaintext string (not a JSON blob). The `secretsmanager:GetSecretValue` permission is required. For quick testing, you can instead use `apiKey` and set the Datadog API key in plaintext.
    - Replace `<FORWARDER_ARN>` with the ARN of your Datadog Lambda Forwarder, as noted previously.

    For additional settings, see [Datadog Serverless Framework Plugin - Configuration parameters][7].

> 5. Enable tracing on your Step Functions by setting the environment variable `DD_TRACE_ENABLED` to `true`.
> 6. For Node.JS and Python runtimes, you can go also [link your step functions with Lambda traces](#associated-lambda-tracing) with associated Lambda tracing.

Reviewer 4: ```suggestion
5. For Node.JS and Python runtimes, set `mergeStepFunctionAndLambdaTraces:true` in your `serverless.yaml` file.
```

[1]: https://docs.datadoghq.com/serverless/libraries_integrations/plugin/
[2]: /logs/guide/forwarder/
[3]: https://app.datadoghq.com/organization-settings/api-keys
[4]: https://www.serverless.com/
[5]: /logs/guide/forwarder/?tab=cloudformation#upgrade-to-a-new-version
[6]: logs/guide/forwarder/?tab=cloudformation#installation
[7]: serverless/libraries_integrations/plugin/#configuration-parameters
{{% /tab %}}
{{% tab "Datadog Serverless CLI" %}}

> 1. If you have not already, install the [Datadog Serverless CLI][1].

Reviewer 4: call out version requirement (2.18.0)

   ```shell
   npm install -g @datadog/datadog-ci
   ```
2. Ensure you have deployed the [Datadog Lambda Forwarder][2], a Lambda function that ships logs from AWS to Datadog, and that you are using v3.66.0+. You may need to [update your Forwarder][3].

   Take note of your Forwarder's ARN.
3. Enable all logging for your Step Function. In your AWS console, open your state machine. Click *Edit* and find the Logging section. There, set *Log level* to `ALL` and enable the *Include execution data* checkbox.

> 4. Enable tracing on your Step Function by setting the environment variable `DD_TRACE_ENABLED` to `true`.

Reviewer 4: ```suggestion
```

> 5. Instrument your Step Function.

Reviewer 4: ```suggestion
4. Instrument your Step Function.
```

   ```shell
   datadog-ci stepfunctions instrument --step-function <STEP_FUNCTION_ARN> --forwarder <FORWARDER_ARN>
   ```
   - Replace `<STEP_FUNCTION_ARN>` with the ARN of your Step Function.
   - Replace `<FORWARDER_ARN>` with the ARN of your Datadog Lambda Forwarder, as noted previously.

   For more information about the `datadog-ci stepfunctions` command, see the [Datadog CLI documentation][5].

> 6. For Node.JS and Python runtimes, you can go also [link your step functions with Lambda traces](#associated-lambda-tracing) with associated Lambda tracing.

Reviewer 4: ```suggestion
5. For Node.JS and Python runtimes, add the flag `-- mergeStepFunctionAndLambdaTraces` in your command
```
